src/malli/core.cljc

```diff
@@ -1958,7 +1958,7 @@
 
 (defn schema?
   "Checks if x is a Schema instance"
-  [x] (#?(:clj instance?, :cljs implements?) malli.core.Schema x))
+  [x] (satisfies? Schema x))
```

Comment:
fast & extendable:

```clj
(defn schema?
  "Checks if x is a Schema instance"
  [x] (or (#?(:clj instance?, :cljs implements?) malli.core.Schema x)) (satisfies? Schema x))
```
  
```

Reply:
Seems pretty neat! Thanks.

However this only results for what we can call the "green path" e.g. `x` was a `malli.core.Schema`.

If `x` belonged an unrelated type e.g. `42`, then the cost of both `or` branches would be inflicted, for nothing.

One possible fix would be to reshuffle the couple `cond`s that use `schema?`:

https://github.com/metosin/malli/blob/68bb521c1112d65b65f6f1276878915169589cdc/src/malli/core.cljc#L1974-L1976

By placing the `schema?` check after the `vector?` check, one surely most code paths won't ever hit the `satisifes?`.

Reply:
Another option would be to make the extra check dependent on a compile-time condition:

```clj
(defmacro maybe-slow-schema-check [x]
  (if (System/getProperty "malli.maybe-slow-schema-check")
    `(satisfies? Schema ~x)
    false))

(defn schema?
  "Checks if x is a Schema instance"
  [x] (or (#?(:clj instance?, :cljs implements?) malli.core.Schema x)) (maybe-slow-schema-check x))
```

This would work with my use case since I only use protocol extensions within a specific dev-time task - not as a general assumption in my production codebase.

Of course it's not as general of a solution but it might be a worthwhile alternative.

Reply:
Thanks for the suggestions. 

There is `malli.perf.creation-perf-test` which has utilities for testing the perf gains / losses.

This code:

```clj
(defn validator []
  (m/validator
   [:map
    [:x boolean?]
    [:y {:optional true} int?]
    [:z [:map
         [:x boolean?]
         [:y {:optional true} int?]]]]))
```

currently calls the `schema?` and `into-schema?` with the args that don't match the predicate:

```clj
"schema?" [:map [:x #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]] [:y {:optional true} #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]] [:z [:map [:x #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]] [:y {:optional true} #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]]]]]
"into-schema?" [:map [:x #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]] [:y {:optional true} #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]] [:z [:map [:x #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]] [:y {:optional true} #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]]]]]
"into-schema?" :map
"schema?" #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]
"into-schema?" #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]
"schema?" #IntoSchema{:type boolean?}
"schema?" #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]
"into-schema?" #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]
"schema?" #IntoSchema{:type int?}
"schema?" [:map [:x #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]] [:y {:optional true} #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]]]
"into-schema?" [:map [:x #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]] [:y {:optional true} #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]]]
"into-schema?" :map
"schema?" #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]
"into-schema?" #object[clojure.core$boolean_QMARK_ 0x53667cbe "clojure.core$boolean_QMARK_@53667cbe"]
"schema?" #IntoSchema{:type boolean?}
"schema?" #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]
"into-schema?" #object[clojure.core$int_QMARK_ 0x21ca93f9 "clojure.core$int_QMARK_@21ca93f9"]
"schema?" #IntoSchema{:type int?}
```

perf:

```clj
(p/bench (validator))
```

* currently: 1.3µs
* just `satisfies?`: 90µs
* or: 90µs
* or + `vector?` check first: 50µs

to make it fast and extendable, we would have implement `Schema` and `IntoSchema` for the default case and use some tagging methods to actually tell if it satisfies the protocol or not. [Sieppari](https://github.com/metosin/sieppari) uses that.  Malli had an elegant hack for caching the `satisifies?` call results, hinted by Alex. Could be re-introduced. Or some other way to write extensions for classes.

lot of work that could be easily fixed in `clojure.core`, go vote the issue up :)

Reply:
> go vote the issue up

Thumbed!

Hope some sort of fix can make it into Malli in the meantime. Feel free to close this PR if you think the final approach would look too different (I'm not familiar with the tagging technique at all).

For now I'll use some monkey patching but I'd sure love something that felt more dependable.